### DISCUSSION TITLE
158 Feature Request: deprecate zen_image_button in admin

### DISCUSSION THREAD

**Discussion by scottcwilson** [zencart/zencart#5780] [Ideas]
@Zen4All did a great job in 1.5.6 in removing calls to this function (reduced from 86 files calling this function to 19 files). 

As of 1.5.7, 13 files still call `zen_image_button` instead of using CSS buttons: 
./attributes_controller.php
./salemaker.php
./gv_queue.php
./includes/functions/html_output.php
./product_types.php
./developers_tool_kit.php
./coupon_admin_export.php
./downloads_manager.php
./media_manager.php
./reviews.php
./coupon_admin.php
./customers.php
./paypal.php

Can we finish the job?

**Comment by Zen4All** [comment]
I can update those files for Zen Cart 1.5.8

For 1.5.7, there is not enough time for me to do the update, and testing.

**Comment by scottcwilson** [comment]
Fantastic!  Great news, thank you.

**Comment by Zen4All** [comment]
@scottcwilson Can you assign me to the task, so I won't forget?

**Comment by scottcwilson** [comment]
I don't have the pow-ah. :(  @drbyte ^^^

**Comment by scottcwilson** [comment]
It would be good to hit the old zen_image_submit calls too if possible.

**Comment by mc12345678** [comment]
So are functions that provide repeating HTML code going away? There seems to be soooo much html in all of the pages now that making a minor adjustment has the potential to have a major impact on the result and there seems to be less "reusable" code with it all being "on page"...  Further, it seems like another "facelift" is going to take even more effort than before to achieve... Am I missing something?

**Comment by drbyte** [comment]
I share a similar concern. We could go either way.

There's strong argument for avoiding use of functions that generate opinionated HTML (not just buttons).

A change of CSS platforms, (even BS3 to BS4) would require more than just an update to buttons, so it's a little weak to argue for buttons to be styled by function just for that sake.

Granted, for plugin support there's some argument for benefit of buttons or other things.

**Comment by drbyte** [comment]
@scottcwilson why the drive for mass exodus from zen_image_submit? and zen_image_button?

**Comment by scottcwilson** [comment]
I think the buttons should look consistent across the admin.  It just looks weird that the heavily used screens like Orders have modern attractive CSS buttons, and the seldom visited screens like Coupon Admin have the old crufty gif buttons.

**Comment by scottcwilson** [comment]
> So are functions that provide repeating HTML code going away?

The scope of this request is just the buttons in admin.  Let's modernize the buttons in admin.  The lion's share was done in 1.5.6; a bit more was done in 1.5.7, let's finish them off in 1.5.8.

**Comment by drbyte** [comment]
> I think the buttons should look consistent across the admin. It just looks weird that the heavily used screens like Orders have modern attractive CSS buttons, and the seldom visited screens like Coupon Admin have the old crufty gif buttons.

I share the desire for consistency. But are we doing it the right way?
I'm surprised you're not advocating to have the functions updated instead ...

**Comment by scottcwilson** [comment]
I don't want to specify methods - that's a decision for the implementer.  I just want the final result of having consistent buttons.

**Comment by Zen4All** [comment]
Can't we rewrite the button functions?
Replacing the new hardcoded buttons with a new function is not that much work.

**Comment by drbyte** [comment]
> Can't we rewrite the button functions?
> Replacing the new hardcoded buttons with a new function is not that much work.

If there are no ugly side-effects to that, I'd welcome it. 
But, as Delia discovered, changing the catalog-side stuff isn't pretty.

Will it cause trouble to plugins that use the functions?

I'd welcome a PR with proof-of-concept on a few pages before wholesale implementation on all the pages.

**Comment by mc12345678** [comment]
Overall, I'd kind of like to see something along the lines of the `table_block`, `box`, or some derivative class that supports the new style formatting and then if/as necessary some sort of new function in lieu of updating the existing. Though compatibility may be better gained by reuse/repurposing of the existing...

**Comment by Zen4All** [comment]
I wrote a new button function, specifically for bootstrap buttons.
This way we will not break existing code
```php
/**
 * Returns the html code for bootsttrap 3 buttons
 * @param string $type The button type: submit, button
 * @param string $text The text on the button: Submit, Cancel, Save
 * @param string $class The Button color: default, primary, succes, info, warning, danger, link
 * @param string $size The button size: xs, sm, lg For normal size leave empty
 * @param string $parameters Things like an id, or a name
 * @return string The html button code
 */
function zen_bootstrap_3_button($type = 'button', $text = '', $class = 'default', $size = '', $parameters = '')
{
  $button = '<button type="' . $type . '" class="btn btn-' . $class . '';
  if (zen_not_null($size)) {
    $button .= ' btn-' . $size;
  }
          $button .= '"';
  if (zen_not_null($parameters)) {
    $button .= ' ' . $parameters;
  }
  $button .= '>' . $text . '</button>';
  return $button;
}
```

**Comment by drbyte** [comment]
@Zen4All 
- should run `zen_output_string_protected` on all the strings that you receive, instead of outputting directly.
- I'm a bit reluctant to name the function after a vendor-specific package. How about `zen_html_button` instead?

(And, BS3 vs BS4 button syntax is the same as far as the way this is implemented.)

**Comment by mc12345678** [comment]
About this:
> should run zen_output_string_protected on all the strings that you receive, instead of outputting directly.

As more abruptly pointed out by the "combination" of admin and catalog "common" functions, this function (once zen_output_string is called) operates differently on the two sides... while would like to see an explanation of why (because it doesn't make sense to me), the real question here is does using the admin's version of the function provide the result that is desired as compared to directly using `htmlspecialchars` and forcing double encoding of the result?

**Comment by Zen4All** [comment]
@drbyte 
- I have not looked at sanitazing inputs for this
- Will change the function

@mc12345678  I'll see what happens with double encoding, and make a choice which sanitazing to use.

**Comment by mc12345678** [comment]
Looking briefly over the function, shouldn't the class assignment consider or handle the case where `empty($class) === true` or at least if `$class` is a set of empty quotes?

**Comment by mc12345678** [comment]
Another question: buttons and other fields have had something that indicates that the button should be hidden based on resolution (identified in the CSS to an extent), could something about that be reasonable to include in the notes? If nothing else, I think it would help those that are seeking to make such a button visible again by being in the code and not require as much forum searching/asking... If not directly in that function, perhaps some common reference be provided/linked to help with that aspect?

**Comment by Zen4All** [comment]
```$class``` Always has to be a value, otherwise the class will be ```class="btn btn-"```

The hiding of fields and buttons is done by a wrapper around the button or field, so this does not have to be part of the function

**Comment by lat9** [comment]
A couple of 'organization' type questions:
1. What if multiple functions were created (like today) to split out the 'normal' from the 'submit' button forms.  That would lengthen the function name, but reduce a parameter.
2. There's a bunch of use of the `<a class="btn btn-primary" href="#" role="button">Link</a>` form of a link-type button.  Perhaps add that as a 3rd function?
3. Perhaps create use-type constants, e.g. `define('BTN_SUBMIT', 'btn btn-primary');`, that are used to provide the button-class?  That way, all submit buttons could be changed 'in one swell foop' in a store's customizations.  It would be kind of like the current use of `button_submit.gif`.
4. Remember, when handling the classes, that a `class=` might be supplied in the `$parameters`.
5. I'm guessing that the parameters are going to be changed much more than the size of a button, how about swapping the `$size` and `$parameters` inputs?

**Comment by Zen4All** [comment]
@lat9 
1. I'll leave that decision to @drbyte 
2. I have been thinking about that too, what would be a good place to define those Constants?
3. Good idea
4. Then people should learn not to do that??
5. I also came to to that conclusion, and already did the swap on my test site

**Comment by lat9** [comment]
For the constants, they'd need to be defined using the
```
if (!defined('X')) define('X', 'something);
```
format in an `extra_datafiles\` file since those files are auto-loaded alphabetically.  That way, a store could _sneak in_ an override in a file that sorts alphabetically before that core-file definition.

For the 'not supplying a class in a button', an emphatic **_NO_**.

**Comment by mc12345678** [comment]
Regarding handling of an empty `$class`
> `$class` Always has to be a value, otherwise the class will be class="btn btn-"

Not sure of the applicability of the comment:
> For the 'not supplying a class in a button', an emphatic **NO**.

My point is that if `''` is provided for `$class` that it seems that it should then be set as `'default'`. This somewhat goes along with the order of priority of the parameters... In attempting to provide a `$parameter` because it is "important" seems that `$class` also then has to be defined although there may be no specific need and thus `''` may seem acceptable when it will cause a problem in the display as identified by the result of `class = "btn btn-"` which was why I raised the question...

**Comment by lat9** [comment]
My comment was with respect to providing an _**additional**_ class for a button.  Removing that functionality or not respecting that a `class=` value could be provided in the `$parameters` is a non-starter.

**Comment by mc12345678** [comment]
> My comment was with respect to providing an additional class for a button. Removing that functionality or not respecting that a class= value could be provided in the $parameters is a non-starter.

Wouldn't this be akin to an additional `name` parameter being provided? E.g., there is already a "place" to capture the parameter to be used/contain whatever is desired to be on the parameter. Seems that to support that `$class` modification either the code being supplied to the function call or a notifier/observer in this additional function would need to support such modification.

**Comment by lat9** [comment]
Seriously?  This isn't complicated.  Just add the 'computed' button-class to the `class=` passed in the $parameters and don't create a separate `class=` tag.

```php
if (strpos($parameters, 'class="') !== false) {
    $parameters = str_replace('class="', 'class="' . $button_class . ' ', $parameters);
}
```

**Comment by drbyte** [comment]
One could also pass their extra class strings "in" the $class parameter, since they'll just be appended to the button class.

**Comment by drbyte** [comment]
I'd rather not add "extra_datafiles" for this. I very much prefer that the files in any of the `extra_` dirs are for non-core things. 

(Yes, I'm aware that there are a small number of existing exceptions to this already in core: I'm trying to avoid making that problem worse. Most of those exceptions are for things that are typically optional or intended as pseudo-examples for plugins/extending.)

**Comment by drbyte** [comment]
> * Perhaps create use-type constants, e.g. `define('BTN_SUBMIT', 'btn btn-primary');`, that are used to provide the button-class?  That way, all submit buttons could be changed 'in one swell foop' in a store's customizations.  It would be kind of like the current use of `button_submit.gif`.

Are such constants really necessary? If one is truly changing ALL `btn-primary` to something else, just do it in the stylesheet.

**Comment by lat9** [comment]
OK, I was just thinking that it might be an easier-read on the code if the 'standard' button colorings were a defined value.

Why not, then, just remove that `$class` parameter and require that the function-callers supply the `class=` tag within the function's `$parameters`?

**Comment by Zen4All** [comment]
Since this is for zc 1.5.8 we can think about the possibilities for some time.
In the meantime I will take your comments, and see what I can do with them.

**Comment by Zen4All** [comment]
So I had some thinking time, and came up with the following proposal:
```php
/**
 * Returns the html code for (bootstrap) buttons
 * 
 * @param string $text The text on the button: Submit, Cancel, Save
 * @param array $class The Button color, size, and other classes<br>
  reserved names for btn- prefix : default, primary, succes, info, warning, danger, link, xs, sm, lg
 * @param string $parameters Things like an id, or a name
 * @param string $type The button type: submit, button
 * @return string The html button code
 */
function zen_html_button($text = '', $class = ['default'], $parameters = '', $type = 'button')
{
  $button = '<button type="' . zen_output_string_protected($type) . '" class="btn' . zen_output_string_protected(zen_get_html_classes($class)) . '"';
  if (zen_not_null($parameters)) {
    $button .= ' ' . $parameters;
  }
  $button .= '>' . $text . '</button>';
  return $button;
}

/**
 * Returns the html code for (bootstrap) link buttons
 * 
 * @param string $link
 * @param string $text
 * @param array $class
 * @param string $parameters
 * @return string
 */
function zen_html_link_button($link = '', $text = '', $class = ['default'], $parameters = '')
{
  $button = '<a href="' . $link . '" title="' . $text . '" ' . $parameters . ' class="btn' . zen_output_string_protected(zen_get_html_classes($class)) . '" role="button">' . $text . '</a>';
  return $button;
}

/**
 * 
 * @param array $class_array
 * @return string
 */
function zen_get_html_classes($class_array)
{
  $classList = '';
  $patern = ['default', 'primary', 'succes', 'info', 'warning', 'danger', 'link', 'xs', 'sm', 'lg'];
  foreach ($class_array as $class) {
    if (in_array($class, $patern)) {
      $classList .= ' btn-' . $class;
    } else {
      $classList .= ' ' . $class;
    }
  }
  return $classList;
}
```

**Comment by drbyte** [comment]
@Zen4All these are good. 
I think I'd rename `zen_get_html_classes()` to `zen_parse_button_classes()`.

I think people are going to dislike having to pass an actual array of class names. Do you foresee any problem with something like this? :

```php
function zen_parse_button_classes($classes = 'default')
{
  $classList = '';
  $btnClasses = ['default', 'primary', 'success', 'info', 'warning', 'danger', 'link', 'xs', 'sm', 'lg'];
  foreach (explode(' ', $classes) as $class) {
    if (in_array($class, $btnClasses) && substr($class, 0, 4) != 'btn-') {
      $classList .= ' btn-' . $class;
    } else {
      $classList .= ' ' . $class;
    }
  }
  return $classList;
}
```

Note the correct english spelling of `success`

**Comment by Zen4All** [comment]
I didn't think an array would be a problem, since zen_get_all_get params is also using an array, but I don't see any problem in using a string.
